File: libs/public-url.js
const url = require('url')

module.exports = ctx => {
  const h = ctx.req.headers
  if (h.referer) {
    const u = url.parse(h.referer)
    return `${u.protocol}//${u.host}`
  } else {
    return `${h['x-forwarded-proto'] || 'http'}://${h.host}`
  }
}

File: cmd/ipa-server/route/publicUrl.go
package route

import (
	"fmt"
	"net/http"
	"net/url"

	"github.com/iineva/ipa-server/pkg/common"
)

func PublicURL(ctx *http.Request) string {
	ref := ctx.Header.Get("referer")
	if ref != "" {
		u, _ := url.Parse(ref)
		return fmt.Sprintf("%v://%v", u.Scheme, u.Host)
	}

	xProto := ctx.Header.Get("x-forwarded-proto")
	host := ctx.Header.Get("host")
	return fmt.Sprintf("%v://%v", common.Def(xProto, "http"), host)
}
